arch/arm/core/userspace.S

```diff
@@ -0,0 +1,154 @@
+/*
+ * Userspace and service handler hooks
+ *
+ * Copyright (c) 2017 Linaro Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ */
+
+#include <offsets_short.h>
+#include <toolchain.h>
+#include <linker/sections.h>
+#include <kernel_structs.h>
+#include <arch/cpu.h>
+#include <syscall.h>
+
+_ASM_FILE_PROLOGUE
+
+GTEXT(_arm_userspace_enter)
+GTEXT(_arm_do_syscall)
+GDATA(_kernel)
+
+/* Imports */
+GTEXT(_k_syscall_table)
+
+/**
+ *
+ * User space entry function
+ *
+ * This function is the entry point to user mode from privileged execution.
+ * The conversion is one way, and threads which transition to user mode do
+ * not transition back later, unless they are doing system calls.
+ *
+ */
+SECTION_FUNC(TEXT,_arm_userspace_enter)
+    /* preserve arguments */
+    mov lr,r0
+    push {r1,r2,r3,lr}
+
+    /* disable mpu stack guard to keep us from triggering guard region */
+    bl arm_core_mpu_disable
+    pop {r1,r2,r3,lr}
+
+    /*
+     * calculate end of stack and set SP to it
+     */
+    ldr r0, [sp]
+    ldr ip, [sp,#4]
+    add r0,ip
+
+    msr PSP, r0
+
+    /* push args to stack, pad end with 2 words of 0x0 */
+    sub sp,8
+    push {r1,r2,r3,lr}
+    ldr r3,=0
+    str r3,[sp, #20]
+    str r3,[sp, #24]
+
+    /* clear the stack area to clean out privileged data */
+#ifdef CONFIG_INIT_STACKS
+    ldr r1,=0xaaaaaaaa
+#else
+    ldr r1,=0
+#endif
+    mov r2,ip
+    sub r2,#24		/* don't erase the pushed regs */
+    bl memset
+
+#ifdef CONFIG_MPU_STACK_GUARD
+    /* setup arguments to thread_entry function */
+    ldr r0,=_kernel
+    ldr r0,[r0, #_kernel_offset_to_current]
+    bl configure_mpu_stack_guard
+    pop {r0,r1,r2,lr}
+#endif
+
+    /* change processor mode to unprivileged */
+    mov ip, #3
+    msr CONTROL, ip
+
+    /* return to thread_entry */
+    bx lr
+
+/**
+ *
+ * Userspace system call function
+ *
+ * This function is used to do system calls from unprivileged code.  This
+ * function is responsible for the following:
+ * 1) Validate the system call.  Must be valid call_id and unprivileged
+ * 2) Switch stacks to the privileged stack
+ * 3) Marshal arguments
+ * 4) Make function call
+ * 5) Clean up by setting back stack and returning
+ *
+ */
+SECTION_FUNC(TEXT, _arm_do_syscall)
+    push {r0,r1,r2,r3,r4,r5,r6,r7}
+    /* validate syscall limit */
+    ldr ip,=_SYSCALL_LIMIT
+    ldr r0, [sp, #40]
```

Comment:
what does this magic value of 40 represent?
can we make this a macro instead?
also this looks very very wrong in that you are validating the system call ID **before** you elevate to privileged mode, which means the checks aren't useful from a security perspective.
you have to do *all* checking after the software interrupt is invoked, nothing done when the CPU is in user mode can be trusted

Reply:
I can move the code to that.  That simplifies some things.